src/inmanta/references.py

```diff
@@ -0,0 +1,467 @@
+        hasher = hashlib.md5()
+        hasher.update(data.encode())
+        return uuid.uuid3(uuid.NAMESPACE_OID, hasher.digest()), arguments
+
+    @property
+    def arguments(self) -> collections.abc.Mapping[str, object]:
+        return {k: v for k, v in self.__dict__.items() if not k.startswith("_")}
+
+
+class Mutator(ReferenceLike):
+    """A mutator that has side effects when executed"""
+
+    @abc.abstractmethod
+    def run(self) -> None:
+        """Execute the mutator"""
+
+    def serialize(self) -> MutatorModel:
+        """Emit the correct pydantic objects to serialize the reference in the exporter."""
+        if not self._model:
+            arg_id, arguments = self.serialize_arguments()
+            self._model = MutatorModel(type=self.type, args=arguments)
+
+        assert isinstance(self._model, MutatorModel)
+        return self._model
+
+
+class Reference[T: RefValue](ReferenceLike):
+    """Instances of this class can create references to a value and resolve them."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._reference_value: T
+        self._reference_value_cached: bool = False
+
+    @classmethod
+    def get_reference_type(cls) -> type[T] | None:
+        # TODO: this will work only when directly extending Reference[T[
+        # TODO: Add tests and docs
+        for g in cls.__orig_bases__:  # type: ignore
+            if typing_inspect.is_generic_type(g) and typing.get_origin(g) in [Reference, DataclassReference]:
+                return typing.get_args(g)[0]
+        return None
+
+    @classmethod
+    def is_dataclass_reference(cls) -> bool:
+        return dataclasses.is_dataclass(cls.get_reference_type())
+
+    @abc.abstractmethod
+    def resolve(self) -> T:
+        """This method resolves the reference and returns the object that it refers to"""
+
+    def get(self) -> T:
+        """Get the value. If we have already resolved it a cached value is returned, otherwise resolve() is called"""
+        if not self._reference_value_cached:
+            self._reference_value = self.resolve()
+            self._reference_value_cached = True
+
+        return self._reference_value
+
+    def serialize(self) -> ReferenceModel:
+        """Emit the correct pydantic objects to serialize the reference in the exporter."""
+        if self._model is CYCLE_TOKEN:
+            raise ReferenceCycleException(self)
+        if not self._model:
+            self._model = CYCLE_TOKEN
+            try:
+                argument_id, arguments = self.serialize_arguments()
+            except ReferenceCycleException as e:
+                e.add(self)
+                raise
+            self._model = ReferenceModel(id=argument_id, type=self.type, args=arguments)
+
+        assert isinstance(self._model, ReferenceModel)
+        return self._model
+
+
+class DataclassReference[T: DataclassProtocol](Reference[T]):
+    pass
```

Comment:
This should get a docstring. What is it used for? Can this even exist?

Reply:
I removed it, it serves no more purpose